### ISSUE TITLE
If an Exception is thrown in a hook, the hook is not used any more after.

### ISSUE THREAD

**Issue by PhilippGrashoff** [atk4/core#84]
Hi there,

```
$model->addHook('beforeSave', function($m) {
    throw new \Exception('some exception');
});

try {
    $model->save();
}
catch(\Exception $e) {
    echo "Exception_found";
}
try {
    $model->save();
}
catch(\Exception $e) {
    echo "Exception_found";
}
```

Expected result: Exception_found is echoed 2 times,
Actual result: echoed only once, beforeSave hook isnt executed any more in second save().

The reason is in \atk4\core\HookTrait in function hook():
A while loop pops the array elements, and, some lines later, added again
```
while ($_data = array_pop($this->hooks[$hook_spot])) {
....
$this->hooks[$hook_spot] = $hook_backup;
```

If an exception is thrown in between these lines, $this->hooks[$hook_spot] is not "filled" again.

A simple fix is to not pop the array elements but leave as is, e.g.
```
try {
            if (
                isset($this->hooks[$hook_spot])
                && is_array($this->hooks[$hook_spot])
            ) {
                krsort($this->hooks[$hook_spot]); // lower priority is called sooner
                $hook_backup = $this->hooks[$hook_spot];
                foreach($this->hooks[$hook_spot] as $_data) {
                    foreach ($_data as &$data) {
                        $return[] = call_user_func_array(
                            $data[0],
                            array_merge(
                                [$this],
                                $arg,
                                $data[1]
                            )
                        );
                    }
                }
            }
        } catch (HookBreaker $e) {
            return $e->return_value;
        }
```

**Comment by PhilippGrashoff** [comment]
Not sure if the popping was for some special reason. If so, after catching the HookBreaker, any other exception could be catched:
```
catch(\Exception $e) {
    $this->hooks[$hook_spot] = $hook_backup;
    throw $e;
}
```

**Comment by DarkSide666** [comment]
Good point.

@romaninsh what you say?

**Comment by romaninsh** [comment]
that popping magic is solving issues when first hook callback adds another hook callback which is yet to be executed. In practice that's done by nested CallbackLater in UI, so any change in logic should be careful about that.

The reason we even handle exceptions here is because we want to trap hook breaking exception. Other exceptions are un-interesting to us and I wanted to leave them alone as much as possible. By default if exception occurs within a while loop, it won't be executed anymore.

In practice - if one hook callback encounters exception (e.g. form's saving), then it would be dangerous to attempt execution of other hooks. 

Here is another illustration:

 1. hook has high priority callback which breaks it.
 2. another callback is there but is never executed.

Now suppose #1 has a bug, and it never reaches a hookBreak point. If the side effect of this bug is that #2 starts executing, then it's dangerous.
